src/Monomer/Graphics/Text.hs

```diff
@@ -413,13 +417,14 @@ isSpaceGroup :: Seq GlyphPos -> Bool
 isSpaceGroup Empty = False
 isSpaceGroup (g :<| gs) = isSpace (_glpGlyph g)
 
-splitGroups :: Seq GlyphPos -> Seq GlyphGroup
-splitGroups Empty = Empty
-splitGroups glyphs = group <| splitGroups rest where
+splitGroups :: LineBreak -> Seq GlyphPos -> Seq GlyphGroup
+splitGroups _ Empty = Empty
+splitGroups break glyphs = group <| splitGroups break rest where
   g :<| gs = glyphs
   groupWordFn = not . isWordDelimiter . _glpGlyph
+  atAnyPos = break == OnCharacters
   (group, rest)
-    | isWordDelimiter (_glpGlyph g) = (Seq.singleton g, gs)
+    | atAnyPos || isWordDelimiter (_glpGlyph g) = (Seq.singleton g, gs)
```

Comment:
This function is slow for a long text with `OnCharacters` like the code I pasted as an example. I don't think this affects the most use cases, though. Should I make this fast?

Reply:
Yes, please!

This is something I was supposed to optimize at some point but, since it worked well enough for my uses cases, I never went back to review it.

Reply:
Sorry I couldn't come up with a good idea.

Reply:
I tried a couple of things, and this looks a bit faster (I have not benchmarked it with Criterion or similar):
```haskell
fitLineToW fontMgr font fSize fSpcH fSpcV metrics break top width trim text = res where
  spaces = T.replicate 4 " "
  newText = T.replace "\t" spaces text
  !glyphs = computeGlyphsPos fontMgr font fSize fSpcH newText
  -- Do not break line on trailing spaces, they are removed in the next step
  -- In the case of KeepSpaces, lines with only spaces (empty looking) are valid
  keepTailSpaces = trim == TrimSpaces
  groups
    | break == OnCharacters = splitGroups break width glyphs
    | otherwise = fitGroups (splitGroups break width glyphs) width keepTailSpaces
  resetGroups
    | trim == TrimSpaces = fmap (resetGlyphs . trimGlyphs) groups
    | otherwise = fmap resetGlyphs groups
  buildLine = buildTextLine font fSize fSpcH fSpcV metrics top
  res
    | text /= "" = Seq.mapWithIndex buildLine resetGroups
    | otherwise = Seq.singleton (buildLine 0 Empty)

...

splitGroups :: LineBreak -> Double -> Seq GlyphPos -> Seq GlyphGroup
splitGroups _ _ Empty = Empty
splitGroups break width glyphs = group <| splitGroups break width rest where
  g :<| gs = glyphs
  groupWordFn = not . isWordDelimiter . _glpGlyph
  groupWidthFn g2 = _glpXMax g2 - _glpXMin g <= width
  atWord = break == OnSpaces
  atAnyPos = break == OnCharacters
  (group, rest)
    | atWord && isWordDelimiter (_glpGlyph g) = (Seq.singleton g, gs)
    | atWord = Seq.spanl groupWordFn glyphs
    | otherwise = Seq.spanl groupWidthFn glyphs
```

Reply:
I checked your code, and I *feel* it's faster, although I haven't run any benchmarks (and TBH, I don't know how to run benchmarks in Haskell).